src/main/java/com/snowflake/kafka/connector/internal/streaming/schemaevolution/iceberg/IcebergTableSchemaResolver.java

```diff
@@ -1,16 +1,133 @@
+      SinkRecord record, Set<String> columnsToEvolve) {
+    if (columnsToEvolve == null || columnsToEvolve.isEmpty()) {
+      return ImmutableList.of();
+    }
+    if (hasSchema(record)) {
+      return getTableSchemaFromRecordSchema(record, columnsToEvolve);
+    } else {
+      return getTableSchemaFromJson(record, columnsToEvolve);
+    }
   }
 
-  public IcebergTableSchemaResolver() {
-    super(new IcebergColumnTypeMapper());
+  private ApacheIcebergColumnSchema mapApacheSchemaFromChannel(
+      Map.Entry<String, ColumnProperties> schemaFromChannelEntry) {
+
+    ColumnProperties columnProperty = schemaFromChannelEntry.getValue();
+    String plainIcebergSchema = getIcebergSchema(columnProperty);
+
+    Type schema = IcebergDataTypeParser.deserializeIcebergType(plainIcebergSchema);
+    String columnName = schemaFromChannelEntry.getKey();
+    return new ApacheIcebergColumnSchema(schema, columnName);
+  }
+
+  // todo remove in 1820155 when getIcebergSchema() method is made public
+  private static String getIcebergSchema(ColumnProperties columnProperties) {
+    try {
+      java.lang.reflect.Field field =
+          columnProperties.getClass().getDeclaredField("icebergColumnSchema");
+      field.setAccessible(true);
+      return (String) field.get(columnProperties);
+    } catch (IllegalAccessException | NoSuchFieldException e) {
+      throw new IllegalStateException(
+          "Couldn't set iceberg by accessing private field: " + "isIceberg", e);
+    }
+  }
+
+  private boolean hasSchema(SinkRecord record) {
+    return record.valueSchema() != null
+        && record.valueSchema().fields() != null
+        && !record.valueSchema().fields().isEmpty();
+  }
+
+  private List<IcebergColumnTree> getTableSchemaFromJson(
+      SinkRecord record, Set<String> columnsToEvolve) {
+    JsonNode recordNode = RecordService.convertToJson(record.valueSchema(), record.value(), true);
+
+    return Streams.stream(recordNode.fields())
+        .map(IcebergColumnJsonValuePair::from)
+        .filter(pair -> columnsToEvolve.contains(pair.getColumnName().toUpperCase()))
+        .map(IcebergColumnTree::new)
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * Given a SinkRecord, get the schema information from it
+   *
+   * @param record the sink record that contains the schema and actual data
+   * @return list of column represantation in a form of tree
+   */
+  private List<IcebergColumnTree> getTableSchemaFromRecordSchema(
+      SinkRecord record, Set<String> columnsToEvolve) {
+
+    Schema schema = record.valueSchema();
+
+    if (schema != null && schema.fields() != null) {
+      ArrayList<Field> foundColumns = new ArrayList<>();
+      ArrayList<Field> notFoundColumns = new ArrayList<>();
+
+      for (Field field : schema.fields()) {
+        if (columnsToEvolve.contains(field.name().toUpperCase())) {
+          foundColumns.add(field);
+        } else {
+          notFoundColumns.add(field);
+        }
+      }
```

Comment:
we could shorten it like this
```suggestion
List<Field> foundColumns = schema.fields().stream()
    .filter(field -> columnsToEvolve.contains(field.name().toUpperCase()))
    .collect(Collectors.toList());

List<Field> notFoundColumns = schema.fields().stream()
    .filter(field -> !columnsToEvolve.contains(field.name().toUpperCase()))
    .collect(Collectors.toList());
```